File: hello_vector_add.cpp
/**
 * TileLang Hello World: Vector Addition (C++ Version)
 * 
 * This example demonstrates using TileLang from C++ to implement
 * element-wise vector addition.
 */

#include <iostream>
#include <vector>
#include <random>
#include <cmath>
#include <chrono>

// TileLang headers would be included here when available
// For now, we'll create a placeholder implementation

void print_banner(const char* message) {
    std::cout << std::string(60, '=') << std::endl;
    std::cout << message << std::endl;
    std::cout << std::string(60, '=') << std::endl;
}

// Simple CPU-based vector addition for demonstration
void vector_add_cpu(const std::vector<float>& a, 
                    const std::vector<float>& b, 
                    std::vector<float>& c) {
    size_t n = a.size();
    for (size_t i = 0; i < n; ++i) {
        c[i] = a[i] + b[i];
    }
}

bool validate_results(const std::vector<float>& result,
                      const std::vector<float>& expected,
                      float tolerance = 1e-5f) {
    if (result.size() != expected.size()) {
        return false;
    }
    
    float max_diff = 0.0f;
    for (size_t i = 0; i < result.size(); ++i) {
        float diff = std::abs(result[i] - expected[i]);
        max_diff = std::max(max_diff, diff);
    }
    
    std::cout << "  Max difference: " << max_diff << std::endl;
    return max_diff < tolerance;
}

int main() {
    print_banner("TileLang Hello World: Vector Addition (C++)");
    
    // Vector size
    const size_t N = 1024 * 1024;  // 1M elements
    
    std::cout << "\nVector size: " << N << " elements" << std::endl;
    std::cout << "Computing: C = A + B (element-wise)" << std::endl;
    
    // Step 1: Prepare test data
    std::cout << "\n[1/4] Preparing test data..." << std::endl;
    
    std::vector<float> a(N), b(N), c(N);
    std::vector<float> expected(N);
    
    // Initialize with random values
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<float> dis(-1.0f, 1.0f);
    
    for (size_t i = 0; i < N; ++i) {
        a[i] = dis(gen);
        b[i] = dis(gen);
    }
    
    std::cout << "✓ Test data prepared" << std::endl;
    
    // Step 2: Compile kernel (placeholder - would use TileLang C++ API)
    std::cout << "\n[2/4] Setting up computation..." << std::endl;
    std::cout << "✓ Using CPU backend (TileLang C++ API integration pending)" << std::endl;
    
    // Step 3: Run computation
    std::cout << "\n[3/4] Running vector addition..." << std::endl;
    
    auto start = std::chrono::high_resolution_clock::now();
    vector_add_cpu(a, b, c);
    auto end = std::chrono::high_resolution_clock::now();
    
    std::chrono::duration<double, std::milli> duration = end - start;
    
    std::cout << "✓ Computation completed" << std::endl;
    
    // Step 4: Validate results
    std::cout << "\n[4/4] Validating results..." << std::endl;
    
    // Create expected output
    vector_add_cpu(a, b, expected);
    
    // Print sample values
    std::cout << "Sample results (first 5 elements):" << std::endl;
    std::cout << "  Output C: [";
    for (int i = 0; i < 5; ++i) {
        std::cout << c[i];
        if (i < 4) std::cout << ", ";
    }
    std::cout << "]" << std::endl;
    
    std::cout << "  Expected: [";
    for (int i = 0; i < 5; ++i) {
        std::cout << expected[i];
        if (i < 4) std::cout << ", ";
    }
    std::cout << "]" << std::endl;
    
    if (validate_results(c, expected)) {
        std::cout << "✓ Results match reference!" << std::endl;
    } else {
        std::cout << "✗ Validation failed" << std::endl;
        return 1;
    }
    
    // Performance metrics
    print_banner("Performance Benchmark");
    
    double latency_ms = duration.count();
    size_t bytes_transferred = N * 3 * sizeof(float);  // 3 vectors * 4 bytes
    double bandwidth_gb_s = (bytes_transferred / 1e9) / (latency_ms / 1000.0);
    
    std::cout << "\nLatency: " << latency_ms << " ms" << std::endl;
    std::cout << "Bandwidth: " << bandwidth_gb_s << " GB/s" << std::endl;
    std::cout << "Throughput: " << (N / (latency_ms * 1e6)) << " GFLOPS" << std::endl;
    
    print_banner("✓ Hello TileLang C++ - Success!");
    
    std::cout << "\nNOTE: This is a CPU reference implementation." << std::endl;
    std::cout << "TileLang C++ API integration is pending for GPU acceleration." << std::endl;
    
    return 0;
}

File: hello_vector_add.py
"""
TileLang Hello World: Vector Addition

This simple example demonstrates the core concepts of TileLang by implementing
element-wise vector addition on GPU/CPU.
"""

import tilelang
import tilelang.language as T


@tilelang.jit
def vector_add(N, block_size=256, dtype="float32"):
    """
    Create a vector addition kernel.
    
    Args:
        N: Size of the vectors
        block_size: Number of threads per block
        dtype: Data type for computation
    """
    
    @T.prim_func
    def vector_add_kernel(
        A: T.Tensor((N,), dtype),
        B: T.Tensor((N,), dtype),
        C: T.Tensor((N,), dtype),
    ):
        # Initialize kernel with grid/block dimensions
        # Each block processes block_size elements
        with T.Kernel(T.ceildiv(N, block_size), threads=block_size) as bx:
            # Allocate thread-local storage for results
            C_local = T.alloc_fragment((block_size,), dtype)
            
            # Parallel computation: each thread handles one element
            for i in T.Parallel(block_size):
                idx = bx * block_size + i
                if idx < N:
                    C_local[i] = A[idx] + B[idx]
            
            # Write results back to global memory
            for i in T.Parallel(block_size):
                idx = bx * block_size + i
                if idx < N:
                    C[idx] = C_local[i]
    
    return vector_add_kernel


def main():
    print("=" * 60)
    print("TileLang Hello World: Vector Addition")
    print("=" * 60)
    
    # Vector size
    N = 1024 * 1024  # 1M elements
    
    print(f"\nVector size: {N:,} elements")
    print(f"Computing: C = A + B (element-wise)")
    
    # Compile the kernel
    print("\n[1/4] Compiling TileLang kernel...")
    kernel = vector_add(N)
    print("✓ Kernel compiled successfully")
    
    # Import torch for testing
    print("\n[2/4] Preparing test data...")
    import torch
    
    # Detect device
    device = "cuda" if torch.cuda.is_available() else "cpu"
    print(f"✓ Using device: {device}")
    
    # Create test vectors
    a = torch.randn(N, device=device, dtype=torch.float32)
    b = torch.randn(N, device=device, dtype=torch.float32)
    c = torch.empty(N, device=device, dtype=torch.float32)
    
    # Run the kernel
    print("\n[3/4] Running TileLang kernel...")
    kernel(a, b, c)
    print("✓ Kernel executed")
    
    # Validate results with simple comparison
    print("\n[4/4] Validating results...")
    ref_c = a + b
    
    # Convert to numpy for comparison
    c_np = c.cpu().numpy()
    ref_np = ref_c.cpu().numpy()
    
    # Calculate maximum absolute difference
    max_diff = abs(c_np - ref_np).max()
    
    # Print sample values
    print(f"Sample results (first 5 elements):")
    print(f"  TileLang output: {c_np[:5]}")
    print(f"  Expected (A+B):  {ref_np[:5]}")
    print(f"  Max difference:  {max_diff:.2e}")
    
    # Check if results are correct (tolerance: 1e-5)
    if max_diff < 1e-5:
        print("✓ Results match reference!")
    else:
        print(f"✗ Validation failed: max difference {max_diff:.2e} exceeds tolerance")
        return
    
    # Benchmark performance
    print("\n" + "=" * 60)
    print("Performance Benchmark")
    print("=" * 60)
    
    profiler = kernel.get_profiler(
        tensor_supply_type=tilelang.TensorSupplyType.Normal
    )
    latency = profiler.do_bench()
    
    # Calculate throughput
    bytes_transferred = N * 3 * 4  # 3 vectors * 4 bytes (float32)
    bandwidth_gb_s = (bytes_transferred / 1e9) / (latency / 1000)
    
    print(f"\nLatency: {latency:.3f} ms")
    print(f"Bandwidth: {bandwidth_gb_s:.2f} GB/s")
    print(f"Throughput: {N / (latency * 1e6):.2f} GFLOPS")
    
    print("\n" + "=" * 60)
    print("✓ Hello TileLang - Success!")
    print("=" * 60)


if __name__ == "__main__":
    main()
